File: weather.py

#!/usr/bin/env python3
# weather.py
import argparse
import json
import csv
import sys
import time
import requests
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class Weather:
    def __init__(self, city, hours=12, interval=0, output_format="text", output_file=None):
        self.city = city
        self.hours = hours
        self.interval = interval
        self.format = output_format
        self.output_file = output_file
        self.data = None

    def fetch(self):
        url = f"https://wttr.in/{self.city}?format=j1&lang=ru"
        try:
            resp = requests.get(url, timeout=10)
            resp.raise_for_status()
            return resp.json()
        except Exception as e:
            print(Fore.RED + f"Ошибка получения данных: {e}")
            return None

    def parse_hourly(self, data):
        if not data or "weather" not in data:
            return []
        hours = data["weather"][0]["hourly"]
        result = []
        for i, h in enumerate(hours[:self.hours]):
            time_str = h["time"][:2] + ":" + h["time"][2:]
            temp = float(h["tempC"])
            feels_like = float(h["FeelsLikeC"])
            humidity = int(h["humidity"])
            wind = float(h["windspeedKmph"])
            precip = float(h["precipMM"])
            cloud = int(h["cloudcover"])
            result.append({
                "time": time_str,
                "temp": temp,
                "feels_like": feels_like,
                "humidity": humidity,
                "wind": wind,
                "precip": precip,
                "cloud": cloud
            })
        return result

    def display_text(self, forecast):
        print(Fore.CYAN + f"Почасовой прогноз для {self.city} (на {self.hours} ч):")
        print("Время | Темп. | Ощущ. | Влажн. | Ветер | Осадки | Облачн.")
        print("-" * 60)
        for f in forecast:
            temp = f["temp"]
            color = Fore.BLUE if temp < 0 else Fore.GREEN if temp < 15 else Fore.YELLOW if temp < 25 else Fore.RED
            print(f"{f['time']:5} | {color}{temp:5.1f}°C{Style.RESET_ALL} | {f['feels_like']:5.1f}°C | {f['humidity']:6}% | {f['wind']:5.1f} км/ч | {f['precip']:6.1f} мм | {f['cloud']:6}%")

    def export_json(self, forecast, filename):
        with open(filename, 'w') as f:
            json.dump(forecast, f, indent=2)
        print(Fore.GREEN + f"Экспортировано в {filename} (JSON)")

    def export_csv(self, forecast, filename):
        with open(filename, 'w', newline='') as f:
            writer = csv.DictWriter(f, fieldnames=["time", "temp", "feels_like", "humidity", "wind", "precip", "cloud"])
            writer.writeheader()
            writer.writerows(forecast)
        print(Fore.GREEN + f"Экспортировано в {filename} (CSV)")

    def run(self):
        while True:
            data = self.fetch()
            if not data:
                break
            forecast = self.parse_hourly(data)
            if not forecast:
                print(Fore.RED + "Нет данных прогноза")
                break
            if self.format == "text":
                self.display_text(forecast)
            elif self.format == "json":
                if self.output_file:
                    self.export_json(forecast, self.output_file)
                else:
                    print(json.dumps(forecast, indent=2))
            elif self.format == "csv":
                if self.output_file:
                    self.export_csv(forecast, self.output_file)
                else:
                    print("CSV вывод требует --output")
            if self.interval == 0:
                break
            time.sleep(self.interval)

def main():
    parser = argparse.ArgumentParser(description="Почасовой прогноз погоды")
    parser.add_argument("--city", default="auto", help="Город (по умолчанию auto)")
    parser.add_argument("--hours", type=int, default=12, help="Количество часов")
    parser.add_argument("--interval", type=int, default=0, help="Интервал обновления (сек)")
    parser.add_argument("--format", choices=["text", "json", "csv"], default="text")
    parser.add_argument("--output", help="Файл для сохранения")
    args = parser.parse_args()

    weather = Weather(args.city, args.hours, args.interval, args.format, args.output)
    try:
        weather.run()
    except KeyboardInterrupt:
        print(Fore.YELLOW + "\nПрервано пользователем")

if __name__ == "__main__":
    main()

File: Weather.cs
// Weather.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace Weather
{
    class Program
    {
        static async Task Main(string[] args)
        {
            var opts = ParseArgs(args);
            var weather = new Weather(opts);
            await weather.RunAsync();
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--city": opts.City = args[++i]; break;
                    case "--hours": opts.Hours = int.Parse(args[++i]); break;
                    case "--interval": opts.Interval = int.Parse(args[++i]); break;
                    case "--format": opts.Format = args[++i]; break;
                    case "--output": opts.Output = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string City { get; set; } = "auto";
            public int Hours { get; set; } = 12;
            public int Interval { get; set; } = 0;
            public string Format { get; set; } = "text";
            public string Output { get; set; }
        }

        class HourlyData
        {
            public string Time { get; set; }
            public double Temp { get; set; }
            public double FeelsLike { get; set; }
            public int Humidity { get; set; }
            public double Wind { get; set; }
            public double Precip { get; set; }
            public int Cloud { get; set; }
        }

        class Weather
        {
            private Options opts;
            private HttpClient client = new HttpClient();

            public Weather(Options opts)
            {
                this.opts = opts;
            }

            private async Task<string> FetchAsync()
            {
                string url = $"https://wttr.in/{opts.City}?format=j1&lang=ru";
                var response = await client.GetAsync(url);
                response.EnsureSuccessStatusCode();
                return await response.Content.ReadAsStringAsync();
            }

            private List<HourlyData> ParseHourly(string json)
            {
                using var doc = JsonDocument.Parse(json);
                var root = doc.RootElement;
                if (!root.TryGetProperty("weather", out var weatherArr) || weatherArr.GetArrayLength() == 0)
                    return new List<HourlyData>();
                var weather = weatherArr[0];
                if (!weather.TryGetProperty("hourly", out var hourlyArr))
                    return new List<HourlyData>();
                var result = new List<HourlyData>();
                int limit = Math.Min(opts.Hours, hourlyArr.GetArrayLength());
                for (int i = 0; i < limit; i++)
                {
                    var h = hourlyArr[i];
                    var time = h.GetProperty("time").GetString();
                    result.Add(new HourlyData
                    {
                        Time = time.Substring(0, 2) + ":" + time.Substring(2),
                        Temp = h.GetProperty("tempC").GetDouble(),
                        FeelsLike = h.GetProperty("FeelsLikeC").GetDouble(),
                        Humidity = h.GetProperty("humidity").GetInt32(),
                        Wind = h.GetProperty("windspeedKmph").GetDouble(),
                        Precip = h.GetProperty("precipMM").GetDouble(),
                        Cloud = h.GetProperty("cloudcover").GetInt32()
                    });
                }
                return result;
            }

            private void DisplayText(List<HourlyData> forecast)
            {
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine($"Почасовой прогноз для {opts.City} (на {opts.Hours} ч):");
                Console.ResetColor();
                Console.WriteLine("Время | Темп. | Ощущ. | Влажн. | Ветер | Осадки | Облачн.");
                Console.WriteLine(new string('-', 60));
                foreach (var f in forecast)
                {
                    ConsoleColor color = ConsoleColor.Blue;
                    if (f.Temp >= 0 && f.Temp < 15) color = ConsoleColor.Green;
                    else if (f.Temp >= 15 && f.Temp < 25) color = ConsoleColor.Yellow;
                    else if (f.Temp >= 25) color = ConsoleColor.Red;
                    Console.ForegroundColor = color;
                    Console.Write($"{f.Time,5} | ");
                    Console.Write($"{f.Temp,5:F1}°C");
                    Console.ResetColor();
                    Console.Write($" | {f.FeelsLike,5:F1}°C | {f.Humidity,6}% | {f.Wind,5:F1} км/ч | {f.Precip,6:F1} мм | {f.Cloud,6}%");
                    Console.WriteLine();
                }
            }

            private void ExportJson(List<HourlyData> forecast, string filename)
            {
                string json = JsonSerializer.Serialize(forecast, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filename, json);
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Экспортировано в {filename} (JSON)");
                Console.ResetColor();
            }

            private void ExportCsv(List<HourlyData> forecast, string filename)
            {
                using var sw = new StreamWriter(filename);
                sw.WriteLine("time,temp,feels_like,humidity,wind,precip,cloud");
                foreach (var f in forecast)
                {
                    sw.WriteLine($"{f.Time},{f.Temp:F1},{f.FeelsLike:F1},{f.Humidity},{f.Wind:F1},{f.Precip:F1},{f.Cloud}");
                }
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Экспортировано в {filename} (CSV)");
                Console.ResetColor();
            }

            public async Task RunAsync()
            {
                while (true)
                {
                    try
                    {
                        string json = await FetchAsync();
                        var forecast = ParseHourly(json);
                        if (forecast.Count == 0)
                        {
                            Console.ForegroundColor = ConsoleColor.Red;
                            Console.WriteLine("Нет данных прогноза");
                            Console.ResetColor();
                            break;
                        }
                        if (opts.Format == "text")
                        {
                            DisplayText(forecast);
                        }
                        else if (opts.Format == "json")
                        {
                            if (!string.IsNullOrEmpty(opts.Output))
                                ExportJson(forecast, opts.Output);
                            else
                                Console.WriteLine(JsonSerializer.Serialize(forecast, new JsonSerializerOptions { WriteIndented = true }));
                        }
                        else if (opts.Format == "csv")
                        {
                            if (!string.IsNullOrEmpty(opts.Output))
                                ExportCsv(forecast, opts.Output);
                            else
                                Console.WriteLine("CSV вывод требует --output");
                        }
                        if (opts.Interval == 0) break;
                        await Task.Delay(opts.Interval * 1000);
                    }
                    catch (Exception ex)
                    {
                        Console.ForegroundColor = ConsoleColor.Red;
                        Console.WriteLine($"Ошибка: {ex.Message}");
                        Console.ResetColor();
                        break;
                    }
                }
            }
        }
    }
}
